parser_api/parser_api.md

```diff
@@ -0,0 +1,1064 @@
+    # without this, no puppet
+    require 'puppet'
+
+    # without this no known environment
+    Puppet.initialize_settings
+
+    # node is needed to get a compiler
+    node = Puppet::Node.new('example')
+
+    # compiler is needed to get a scope
+    compiler = Puppet::Parser::Compiler.new(node)
+    scope = compiler.topscope
+
+    # Test that it works
+    puts Puppet::Pops::Parser::EvaluatingParser.evaluate_string(scope, "yay")
+
+We can then run that to see it working:
+
+    > bundle exec ruby test.rb
+    yay
+
+Depending on hour your ruby and puppet is installed you may need to tweak how you get ruby to run it and find puppet.
+Note that the "bare metal" setup uses the settings for Puppet. If you have configured the default "production" environment
+with a module path containing modules, you will be able to call functions in modules as well as in puppet.
+
+Note that evaluation done this way does not actually produce a catalog, even if one is built up in the compiler that
+is created. How to do a full compilation is beyond the scope of this document.
+
+The following examples assumes a similar setup so a `scope` variable is available.
+
+Simple example - adding with puppet:
+
+    parser = Puppet::Pops::Parser::EvaluatingParser.singleton
+    puts parser.evaluate_string(scope, "1 + 1") # prints 2
+    puts parser.evaluate_string(scope, "$a = 10; $a + 20") # prints 30
+
+Parsing
+---
+
+Parsing is just as simple as evaluation:
+
+    parser = Puppet::Pops::Parser::EvaluatingParser.singleton
+    result = parser.parse_string('$a = 1 + 1', 'testing.pp')
+
+With this API, the result that is produced is syntactically and semantically correct (but there may be errors in the logic
+that is not detected until it is evaluated), or you will get an exception and the errors and warnings have been logged - just like when compiling a catalog. The `parse_file` method behave the same way, but reads the content from a file instead of being given a String with the source.
+
+### Parsing EPP
+
+Parsing EPP works the same way as parsing regular puppet. Here `EvaluatingEppParser` is used instead of `EvaluatingParser`.
+Also not that this class does not have a `singleton` method, so it is required to call `new` to get an instance of it.
+
+    parser = Puppet::Pops::Parser::EvaluatingEppParser.new
+    result = parser.parse_string('<%= 1 + 1%>', 'testing.epp')
+
+The result produced works exactly the same way as when parsing regular PP logic.
+
+### The Parse Result
+
+The result of parsing is an instance of `Puppet::Pops::Model::Factory`, on which there is one public method named `model` that returns the Abstract Syntax Tree (AST) that represents the parsed source. The root of this tree is always an instance of
+`Puppet::Pops::Model::Program`. This class, along with the other close to 90 classes that make up all kinds of expressions and statements in the puppet language are found in the file `lib/puppet/pops/model/model_meta.rb`
+
```

Comment:
The above changes in Puppet 5.0.0 so that there is no need to unwrap the returned factory using the call to `model`. We added a backwards compatible method which still allows calling `model` on the return value, that then just returns self.